### DISCUSSION TITLE
How to render a component by user agent

### DISCUSSION THREAD

**Discussion by Capricair** [nuxt/nuxt#10043] [Questions (Nuxt 2)]
Hello

I want render a mobile web page by user agent, I added a module in my project and rewrite renderRoute function, like this

```
export default async function renderPage(moduleOptions) {
  const nuxt = this.nuxt
  const renderer = nuxt.renderer
  const renderRoute = renderer.renderRoute.bind(renderer)

  renderer.renderRoute = async (url, context, retried) => {
    const userAgent = context.req.headers['user-agent']
    const isMobile = /(Android|webOS|iPhone|iPad|iPod|tablet|BlackBerry|Mobile)/i.test(userAgent)
    const component = isMobile ? componentM : componentPC
    return await renderRoute(component, context, retried)
  }
}
```
it didn't work, what should I do, I need some help, thanks a lot!

I know next.js has a renderToHTML function can do this
```
const app = next({ dev: isDevelopment })
const server = express()
const handle = app.getRequestHandler()

server.get('/your/url', async (req, res) => {
  const userAgent = req.headers['user-agent']
  const isMobile =  /(Android|webOS|iPhone|iPad|iPod|tablet|BlackBerry|Mobile)/i.test(userAgent)
  const actualPage = isMobile ? '/mobile/page/path' : '/pc/page/path'
  const html = await app.renderToHTML(req, res, actualPage, req.params)
  res.send(html)
})

server.get('*', (req, res) => {
  handle(req, res)
})
```
Does nuxt have the same method?

PS: For some reasons, I can’t do this by redirect to a new url

**Comment by yanqc1996** [comment]
Have you solved this problem? I encountered a similar problem and want to know how to solve it

**Reply by Capricair** [reply]
No, I sent message to the nuxt project team members on github, contacted them on almost all social media, but got no response. I have to say that the ecology of vue is too bad, we are already using react and next.js

**Reply by devzom** [reply]
@Capricair it's still valid issue for 2.15.8 version ?

**Comment by paulvonber** [comment]
Why don't you use `<component :is="" />` within the page? then you can render whatever you want by matching whatever condition

**Reply by Capricair** [reply]
If you do this, you must import both the PC and mobile, and the packaged files will contain useless css and js, which will lead to longer page loading times.

**Comment by devzom** [comment]
I've resolved a few of cases like this by using > [@nuxtjs/device package](https://github.com/nuxt-community/device-module) You can watch for a specific agents or check if it's a mobile, as it's well preconfigured.
try it out :)

**Reply by Capricair** [reply]
My problem is how to render the page by userAgent, not how to detect the device type

**Reply by devzom** [reply]
I see :) but You can also preconfigure a  [custom userAgent](https://github.com/nuxt-community/device-module#add-a-custom-flag)

**Reply by Capricair** [reply]
Thanks, but my problem has nothing to do with this
